### PR TITLE
Run wrench on travis CI with OSMesa

### PR DESCRIPTION
**Pull request by glennw**
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/649)

### CODE REVIEW

File: wrench/Cargo.toml [hunk 1]
```diff
@@ -33,3 +33,7 @@ dwrote = "0.1.2"
 
 [target.'cfg(target_os = "linux")'.dependencies]
 font-loader = "0.1.0"
+
+[target.'cfg(any(target_os = "linux", target_os = "macos"))'.dependencies]
```
Comment (kvark) [inline wrench/Cargo.toml]: given that we use it for automated tests only, can we hide this dependency behind a feature gate?

File: wrench/run_tests.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    elif is_macosx():
+        osmesa_path = path.join(find_dep_path_newest('osmesa-src', bin_path),
+                                "out", "src", "gallium", "targets", "osmesa", ".libs")
+        glapi_path = path.join(find_dep_path_newest('osmesa-src', bin_path),
+                               "out", "src", "mapi", "shared-glapi", ".libs")
+        os.environ["DYLD_LIBRARY_PATH"] = osmesa_path + ":" + glapi_path
+        os.environ["GALLIUM_DRIVER"] = "softpipe"
+
+
+set_osmesa_env('../target/release/')
```
Comment (kvark) [inline wrench/run_tests.py]: is our executable built by `cargo test` guaranteed to be in `release` folder?

File: wrench/src/main.rs [hunk 3]
```diff
@@ -75,31 +86,161 @@ impl ThingKind {
...
+            width: width,
+            height: height,
+        }
+    }
+
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
+    fn get_proc_address(s: &str) -> *const c_void {
+        let c_str = CString::new(s).expect("Unable to create CString");
+        unsafe {
+            mem::transmute(osmesa_sys::OSMesaGetProcAddress(c_str.as_ptr()))
```
Comment (kvark) [inline wrench/src/main.rs]: can we use a simple `X as *const _` here instead of `transmute`?

File: wrench/src/main.rs [hunk 4]
```diff
@@ -175,97 +320,138 @@ fn main() {
...
+            if let Some(limit) = limit_seconds {
+                if (time::SteadyTime::now() - time_start) >= limit {
+                    break 'outer;
                 }
+            }
 
-                if show_help {
-                    wrench.show_onscreen_help();
-                }
+            let thing = thing.thing();
```
Comment (kvark) [inline wrench/src/main.rs]: thing

File: wrench/src/main.rs [hunk 5]
```diff
@@ -175,97 +320,138 @@ fn main() {
...
-                            VirtualKeyCode::Right => {
-                                thing.next_frame();
-                            },
-                            VirtualKeyCode::H => {
-                                show_help = !show_help;
-                            },
-                            _ => ()
+                glutin::Event::KeyboardInput(kstate, _scan_code, maybe_vk) => {
+                    if kstate == ElementState::Pressed {
+                        if let Some(vk) = maybe_vk {
```
Comment (kvark) [inline wrench/src/main.rs]: nit: we could combine both of these `if`s

File: wrench/src/main.rs [hunk 6]
```diff
@@ -175,97 +320,138 @@ fn main() {
...
             }
-            _ => ()
         }
     }
+
+    if is_headless {
+        let mut pixels = gl::read_pixels(0, 0,
+                                         size.width as gl::GLsizei,
+                                         size.height as gl::GLsizei,
+                                         gl::RGB, gl::UNSIGNED_BYTE);
```
Comment (kvark) [inline wrench/src/main.rs]: would that incur a software format conversion from RGBA to RGB?

**Comment by glennw** [comment]
This is a bit untidy, but it now runs wrench in headless / osmesa mode on travis (mac + linux), writes a screenshot, and the md5sum of that png matches on both CI machines, and also on my local machine when I check the image output visually.

@jrmuizel Hopefully this is enough for you to continue your work on adding basic test support. Let me know if anything isn't working or needs some additional functionality.

@kvark I've addressed most of the comments you left, but if you have time to take another look over it and see what else needs cleaning up, we could perhaps get this part merged soon-ish.

File: wrench/Cargo.toml (hunk 1 above)
Comment (glennw) [inline wrench/Cargo.toml]: Moved this to the 'headless' feature.

File: wrench/run_tests.py (hunk 2 above)
Comment (glennw) [inline wrench/run_tests.py]: I explicitly build the release version in the travis script - perhaps there's a better way to do this though.

File: .travis.yml [hunk 7]
```diff
@@ -8,8 +8,15 @@ notifications:
   webhooks: http://build.servo.org:54856/travis
 addons:
   apt:
+    sources:
+      - sourceline: 'deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.8 main'
+        keyurl: 'http://apt.llvm.org/llvm-snapshot.gpg.key'
     packages:
       - libgles2-mesa-dev
+      - llvm-3.8
+      - llvm-3.8-dev
```
Comment (glennw) [inline .travis.yml]: Fixed

File: wrench/src/main.rs [hunk 8]
```diff
@@ -21,18 +21,29 @@ extern crate serde;
 extern crate serde_json;
 extern crate crossbeam;
 
+#[cfg(any(target_os = "linux", target_os = "macos"))]
```
Comment (glennw) [inline wrench/src/main.rs]: Added the headless feature.

File: wrench/src/main.rs (hunk 3 above)
Comment (glennw) [inline wrench/src/main.rs]: The OSMesa bindings return an Option<fn> here so I think a transmute is required here due to the interface mismatch. We could perhaps make these match in the future and remove the transmute?

File: wrench/src/main.rs (hunk 5 above)
Comment (glennw) [inline wrench/src/main.rs]: Unrelated to this PR, so leaving for now.

File: wrench/src/main.rs (hunk 6 above)
Comment (glennw) [inline wrench/src/main.rs]: Perhaps - I'm not entirely sure. But it's only done once when writing the screenshot in test mode, so probably not a big deal.

**Comment by nox** [comment]
@glennw Seems that Travis is happy now.

File: wrench/src/main.rs (hunk 3 above)
Comment (kvark) [inline wrench/src/main.rs]: Sounds good! No need to have an extra `unsafe` where it's not required

File: wrench/src/main.rs [hunk 9]
```diff
@@ -21,18 +21,29 @@ extern crate serde;
...
 use std::path::PathBuf;
 use std::cmp::{min, max};
-use std::ffi::CStr;
+use std::fs::File;
+use std::mem;
+use std::os::raw::c_void;
+use std::ptr;
 use webrender_traits::*;
 
+#[cfg(any(target_os = "linux", target_os = "macos"))]
```
Comment (kvark) [inline wrench/src/main.rs]: should use `headless` feature

File: wrench/src/main.rs [hunk 10]
```diff
@@ -75,31 +86,161 @@ impl ThingKind {
...
+}
+
+pub struct HeadlessEventIterater;
+
+impl WindowWrapper {
+    fn init(&self) {
+        match *self {
+            WindowWrapper::Window(ref window) => {
+                unsafe {
+                    window.make_current().ok();
```
Comment (kvark) [inline wrench/src/main.rs]: I don't think it makes sense to ignore the returned `Err` here, we better just do `unwrap` or `expect` instead.

File: wrench/src/main.rs [hunk 11]
```diff
@@ -175,97 +320,138 @@ fn main() {
...
+        let mut events = Vec::new();
+
+        match window {
+            WindowWrapper::Headless(..) => {
+                // Just keep rendering!
+                // TODO: Add an exit condition for headless mode!
+                events.push(glutin::Event::Awakened);
+            }
+            WindowWrapper::Window(ref window) => {
+                for event in window.poll_events() {
```
Comment (kvark) [inline wrench/src/main.rs]: nit: could do shorter:
```rust
let events = match window {
  WindowWrapper::Headless(..) => {
    vec![glutin::Event::Awakened]
  }
  WindowWrapper::Window(ref window) => {
    window.poll_events().collect()
  }
};
```

**Comment by glennw** [comment]
@kvark Thanks for the review, addressed those comments and squashed.

**Comment by kvark** [comment]
Thanks!
I suppose this error at the end is expected?
> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libcore/result.rs:837
note: Run with `RUST_BACKTRACE=1` for a backtrace.
md5 = 31ca82c0aadec2c0ebfd0e3f95510c9c
/Users/travis/build.sh: line 57: shell_session_update: command not found

Either way, we can follow up with more work, at least Travis is green.
@bors-servo r+

**Pull request merged**